File: starter-kits/adk-python/shared/wiki_lookup_tool.py
#!/usr/bin/env python3
"""Minimal wiki lookup helper for ADK runtime context building.

This helper reads the local LLM wiki, finds a small set of relevant pages,
and returns a compact evidence bundle that a Wiki Knowledge Enricher agent can
turn into ``learning_context``.
"""

from __future__ import annotations

import argparse
import json
import re
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Iterable, Sequence

EXCLUDED_FILE_NAMES = {"README.md", "index.md", "log.md"}
EXCLUDED_DIRECTORIES = {"lint", "queries", "source-notes"}
TOKEN_PATTERN = re.compile(r"[A-Za-z]{2,}|[가-힣]{2,}")


@dataclass
class WikiMatch:
    path: str
    score: float
    excerpt: str


@dataclass
class WikiContext:
    target_profile: str
    wiki_root: str
    query_keywords: list[str]
    index_excerpt: list[str]
    matches: list[WikiMatch]
    usage_notes: list[str]


def resolve_wiki_root(start_path: str | Path | None = None) -> Path:
    """Resolve the local wiki directory from cwd or a provided hint path."""

    start = Path(start_path).resolve() if start_path else Path.cwd().resolve()
    current = start if start.is_dir() else start.parent

    while True:
        direct_wiki = current / "wiki"
        if (direct_wiki / "index.md").exists():
            return direct_wiki

        if current.name == "wiki" and (current / "index.md").exists():
            return current

        if current.parent == current:
            break
        current = current.parent

    raise FileNotFoundError("Could not locate wiki/index.md from the provided path.")


def _strip_frontmatter(text: str) -> str:
    lines = text.splitlines()
    if not lines or lines[0].strip() != "---":
        return text

    for index, line in enumerate(lines[1:], start=1):
        if line.strip() == "---":
            return "\n".join(lines[index + 1 :]).strip()

    return text


def _normalize_keywords(
    problem_text: str, approved_solution_summary: str, concept_candidates: Sequence[str]
) -> list[str]:
    keywords: list[str] = []
    seen: set[str] = set()

    def add_keyword(raw_value: str) -> None:
        value = raw_value.strip()
        key = value.casefold()
        if not value or key in seen:
            return
        if value.isdigit():
            return
        seen.add(key)
        keywords.append(value)

    for concept in concept_candidates:
        add_keyword(concept)

    for token in TOKEN_PATTERN.findall(f"{problem_text} {approved_solution_summary}"):
        add_keyword(token)
        if len(keywords) >= 12:
            break

    return keywords


def _relative_page_path(page_path: Path, wiki_root: Path) -> str:
    repo_root = wiki_root.parent
    return str(page_path.relative_to(repo_root))


def _collect_index_excerpt(index_text: str, keywords: Sequence[str], limit: int = 6) -> list[str]:
    excerpts: list[str] = []
    lowered_keywords = [keyword.casefold() for keyword in keywords]

    for raw_line in index_text.splitlines():
        line = raw_line.strip()
        if not line:
            continue
        lowered_line = line.casefold()
        if any(keyword in lowered_line for keyword in lowered_keywords):
            excerpts.append(line)
        if len(excerpts) >= limit:
            break

    return excerpts


def _extract_excerpt(body_text: str, keywords: Sequence[str], limit: int = 280) -> str:
    lowered_keywords = [keyword.casefold() for keyword in keywords]
    lines = [line.strip() for line in body_text.splitlines() if line.strip()]

    for index, line in enumerate(lines):
        lowered_line = line.casefold()
        if any(keyword in lowered_line for keyword in lowered_keywords):
            if line.startswith("#") and index + 1 < len(lines):
                candidate = f"{line} {lines[index + 1]}"
                return candidate[:limit]
            return line[:limit]

    for line in lines:
        if not line.startswith("#"):
            return line[:limit]

    return ""


def _score_page(page_path: Path, body_text: str, keywords: Sequence[str]) -> float:
    lowered_body = body_text.casefold()
    lowered_name = page_path.stem.replace("-", " ").casefold()
    path_parts = set(page_path.parts)
    score = 0.0

    for keyword in keywords:
        lowered_keyword = keyword.casefold()
        match_count = min(lowered_body.count(lowered_keyword), 5)
        score += float(match_count)
        if lowered_keyword in lowered_name:
            score += 3.0

    if "syntheses" in path_parts:
        score += 2.0
    if "components" in path_parts or "overview" in path_parts:
        score += 0.5

    return score


def _iter_candidate_pages(wiki_root: Path) -> Iterable[Path]:
    for page_path in wiki_root.rglob("*.md"):
        if page_path.name in EXCLUDED_FILE_NAMES:
            continue
        if any(part in EXCLUDED_DIRECTORIES for part in page_path.parts):
            continue
        yield page_path


def lookup_wiki_context(
    problem_text: str,
    concept_candidates: Sequence[str] | None = None,
    target_profile: str = "KR-2022-middle",
    approved_solution_summary: str = "",
    max_pages: int = 4,
    wiki_root: str | Path | None = None,
) -> dict:
    """Return a compact evidence bundle for a post-review wiki enrichment step."""

    concept_candidates = concept_candidates or []
    wiki_root_path = resolve_wiki_root(wiki_root)
    keywords = _normalize_keywords(problem_text, approved_solution_summary, concept_candidates)
    index_text = (wiki_root_path / "index.md").read_text(encoding="utf-8")

    scored_matches: list[WikiMatch] = []
    for page_path in _iter_candidate_pages(wiki_root_path):
        raw_text = page_path.read_text(encoding="utf-8")
        body_text = _strip_frontmatter(raw_text)
        score = _score_page(page_path, body_text, keywords)
        if score <= 0:
            continue
        scored_matches.append(
            WikiMatch(
                path=_relative_page_path(page_path, wiki_root_path),
                score=round(score, 2),
                excerpt=_extract_excerpt(body_text, keywords),
            )
        )

    scored_matches.sort(key=lambda item: (-item.score, item.path))
    context = WikiContext(
        target_profile=target_profile,
        wiki_root=str(wiki_root_path),
        query_keywords=keywords,
        index_excerpt=_collect_index_excerpt(index_text, keywords),
        matches=scored_matches[:max_pages],
        usage_notes=[
            "Run this lookup only after reviewer approval.",
            "Use the result to build a compact learning_context, not to guess the math answer.",
            "If the matches are weak or sparse, mark curriculum mapping and difficulty as tentative.",
        ],
    )
    return asdict(context)


def build_learning_context_seed(
    problem_text: str,
    concept_candidates: Sequence[str] | None = None,
    target_profile: str = "KR-2022-middle",
    approved_solution_summary: str = "",
    max_pages: int = 4,
    wiki_root: str | Path | None = None,
) -> dict:
    """Alias for teams that prefer context-builder naming."""

    return lookup_wiki_context(
        problem_text=problem_text,
        concept_candidates=concept_candidates,
        target_profile=target_profile,
        approved_solution_summary=approved_solution_summary,
        max_pages=max_pages,
        wiki_root=wiki_root,
    )


def main() -> None:
    parser = argparse.ArgumentParser(description="Build a compact wiki context seed.")
    parser.add_argument("problem_text", help="Original math problem text.")
    parser.add_argument(
        "--concept",
        action="append",
        default=[],
        help="Concept candidates extracted by the problem interpreter.",
    )
    parser.add_argument(
        "--approved-solution-summary",
        default="",
        help="Optional short summary of the approved solution.",
    )
    parser.add_argument(
        "--target-profile",
        default="KR-2022-middle",
        help="Curriculum profile used for the enrichment step.",
    )
    parser.add_argument(
        "--max-pages",
        type=int,
        default=4,
        help="Maximum number of wiki pages to keep in the result.",
    )
    parser.add_argument(
        "--wiki-root",
        default=None,
        help="Optional path hint to the repo root or the wiki directory.",
    )
    args = parser.parse_args()

    result = lookup_wiki_context(
        problem_text=args.problem_text,
        concept_candidates=args.concept,
        target_profile=args.target_profile,
        approved_solution_summary=args.approved_solution_summary,
        max_pages=args.max_pages,
        wiki_root=args.wiki_root,
    )
    print(json.dumps(result, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: starter-kits/adk-go/shared/wiki_lookup_tool.go
package wikilookup

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var tokenPattern = regexp.MustCompile(`[A-Za-z]{2,}|[가-힣]{2,}`)

var excludedFileNames = map[string]bool{
	"README.md": true,
	"index.md":  true,
	"log.md":    true,
}

var excludedDirectories = map[string]bool{
	"lint":         true,
	"queries":      true,
	"source-notes": true,
}

// WikiMatch is one ranked wiki page candidate for post-review enrichment.
type WikiMatch struct {
	Path    string  `json:"path"`
	Score   float64 `json:"score"`
	Excerpt string  `json:"excerpt"`
}

// WikiContext is a compact evidence bundle that a Wiki Knowledge Enricher
// agent can turn into learning_context.
type WikiContext struct {
	TargetProfile string      `json:"target_profile"`
	WikiRoot      string      `json:"wiki_root"`
	QueryKeywords []string    `json:"query_keywords"`
	IndexExcerpt  []string    `json:"index_excerpt"`
	Matches       []WikiMatch `json:"matches"`
	UsageNotes    []string    `json:"usage_notes"`
}

// ResolveWikiRoot finds the local wiki directory from a starting path.
func ResolveWikiRoot(startPath string) (string, error) {
	base := startPath
	if base == "" {
		cwd, err := os.Getwd()
		if err != nil {
			return "", err
		}
		base = cwd
	}

	absoluteBase, err := filepath.Abs(base)
	if err != nil {
		return "", err
	}

	info, err := os.Stat(absoluteBase)
	if err == nil && !info.IsDir() {
		absoluteBase = filepath.Dir(absoluteBase)
	}

	current := absoluteBase
	for {
		directWiki := filepath.Join(current, "wiki", "index.md")
		if _, err := os.Stat(directWiki); err == nil {
			return filepath.Join(current, "wiki"), nil
		}

		if filepath.Base(current) == "wiki" {
			directIndex := filepath.Join(current, "index.md")
			if _, err := os.Stat(directIndex); err == nil {
				return current, nil
			}
		}

		parent := filepath.Dir(current)
		if parent == current {
			break
		}
		current = parent
	}

	return "", fmt.Errorf("could not locate wiki/index.md from %q", startPath)
}

// LookupWikiContext reads the local wiki and returns a compact evidence bundle.
func LookupWikiContext(
	problemText string,
	approvedSolutionSummary string,
	conceptCandidates []string,
	targetProfile string,
	maxPages int,
	startPath string,
) (WikiContext, error) {
	if maxPages <= 0 {
		maxPages = 4
	}
	if targetProfile == "" {
		targetProfile = "KR-2022-middle"
	}

	wikiRoot, err := ResolveWikiRoot(startPath)
	if err != nil {
		return WikiContext{}, err
	}

	indexBytes, err := os.ReadFile(filepath.Join(wikiRoot, "index.md"))
	if err != nil {
		return WikiContext{}, err
	}

	keywords := normalizeKeywords(problemText, approvedSolutionSummary, conceptCandidates)
	matches, err := collectMatches(wikiRoot, keywords)
	if err != nil {
		return WikiContext{}, err
	}
	if len(matches) > maxPages {
		matches = matches[:maxPages]
	}

	return WikiContext{
		TargetProfile: targetProfile,
		WikiRoot:      wikiRoot,
		QueryKeywords: keywords,
		IndexExcerpt:  collectIndexExcerpt(string(indexBytes), keywords, 6),
		Matches:       matches,
		UsageNotes: []string{
			"Run this lookup only after reviewer approval.",
			"Use the result to build a compact learning_context, not to guess the math answer.",
			"If the matches are weak or sparse, mark curriculum mapping and difficulty as tentative.",
		},
	}, nil
}

func normalizeKeywords(problemText string, approvedSolutionSummary string, conceptCandidates []string) []string {
	keywords := make([]string, 0, 12)
	seen := map[string]bool{}

	addKeyword := func(value string) {
		trimmed := strings.TrimSpace(value)
		lowered := strings.ToLower(trimmed)
		if trimmed == "" || seen[lowered] {
			return
		}
		if isDigitsOnly(trimmed) {
			return
		}
		seen[lowered] = true
		keywords = append(keywords, trimmed)
	}

	for _, concept := range conceptCandidates {
		addKeyword(concept)
	}

	for _, token := range tokenPattern.FindAllString(problemText+" "+approvedSolutionSummary, -1) {
		addKeyword(token)
		if len(keywords) >= 12 {
			break
		}
	}

	return keywords
}

func collectMatches(wikiRoot string, keywords []string) ([]WikiMatch, error) {
	matches := []WikiMatch{}
	repoRoot := filepath.Dir(wikiRoot)

	err := filepath.WalkDir(wikiRoot, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			if excludedDirectories[d.Name()] {
				return filepath.SkipDir
			}
			return nil
		}
		if filepath.Ext(path) != ".md" || excludedFileNames[filepath.Base(path)] {
			return nil
		}

		rawBytes, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		body := stripFrontmatter(string(rawBytes))
		score := scorePage(path, body, keywords)
		if score <= 0 {
			return nil
		}

		relativePath, err := filepath.Rel(repoRoot, path)
		if err != nil {
			return err
		}
		matches = append(matches, WikiMatch{
			Path:    filepath.ToSlash(relativePath),
			Score:   score,
			Excerpt: extractExcerpt(body, keywords, 280),
		})
		return nil
	})
	if err != nil {
		return nil, err
	}

	sort.Slice(matches, func(i, j int) bool {
		if matches[i].Score == matches[j].Score {
			return matches[i].Path < matches[j].Path
		}
		return matches[i].Score > matches[j].Score
	})

	return matches, nil
}

func collectIndexExcerpt(indexText string, keywords []string, limit int) []string {
	excerpts := []string{}
	loweredKeywords := make([]string, 0, len(keywords))
	for _, keyword := range keywords {
		loweredKeywords = append(loweredKeywords, strings.ToLower(keyword))
	}

	for _, rawLine := range strings.Split(indexText, "\n") {
		line := strings.TrimSpace(rawLine)
		if line == "" {
			continue
		}
		loweredLine := strings.ToLower(line)
		for _, keyword := range loweredKeywords {
			if strings.Contains(loweredLine, keyword) {
				excerpts = append(excerpts, line)
				break
			}
		}
		if len(excerpts) >= limit {
			break
		}
	}

	return excerpts
}

func stripFrontmatter(text string) string {
	lines := strings.Split(text, "\n")
	if len(lines) == 0 || strings.TrimSpace(lines[0]) != "---" {
		return text
	}

	for index := 1; index < len(lines); index++ {
		if strings.TrimSpace(lines[index]) == "---" {
			return strings.TrimSpace(strings.Join(lines[index+1:], "\n"))
		}
	}

	return text
}

func scorePage(path string, body string, keywords []string) float64 {
	loweredBody := strings.ToLower(body)
	loweredName := strings.ToLower(strings.ReplaceAll(strings.TrimSuffix(filepath.Base(path), filepath.Ext(path)), "-", " "))
	pathParts := strings.Split(filepath.ToSlash(path), "/")
	pathSet := map[string]bool{}
	for _, part := range pathParts {
		pathSet[part] = true
	}

	score := 0.0
	for _, keyword := range keywords {
		loweredKeyword := strings.ToLower(keyword)
		score += float64(min(strings.Count(loweredBody, loweredKeyword), 5))
		if strings.Contains(loweredName, loweredKeyword) {
			score += 3.0
		}
	}

	if pathSet["syntheses"] {
		score += 2.0
	}
	if pathSet["components"] || pathSet["overview"] {
		score += 0.5
	}

	return score
}

func extractExcerpt(body string, keywords []string, limit int) string {
	loweredKeywords := make([]string, 0, len(keywords))
	for _, keyword := range keywords {
		loweredKeywords = append(loweredKeywords, strings.ToLower(keyword))
	}

	lines := []string{}
	for _, rawLine := range strings.Split(body, "\n") {
		line := strings.TrimSpace(rawLine)
		if line != "" {
			lines = append(lines, line)
		}
	}

	for index, line := range lines {
		loweredLine := strings.ToLower(line)
		for _, keyword := range loweredKeywords {
			if strings.Contains(loweredLine, keyword) {
				if strings.HasPrefix(line, "#") && index+1 < len(lines) {
					return trimToLimit(line+" "+lines[index+1], limit)
				}
				return trimToLimit(line, limit)
			}
		}
	}

	for _, line := range lines {
		if !strings.HasPrefix(line, "#") {
			return trimToLimit(line, limit)
		}
	}

	return ""
}

func trimToLimit(value string, limit int) string {
	if limit <= 0 {
		return value
	}
	runes := []rune(value)
	if len(runes) <= limit {
		return value
	}
	return string(runes[:limit])
}

func isDigitsOnly(value string) bool {
	for _, char := range value {
		if char < '0' || char > '9' {
			return false
		}
	}
	return value != ""
}

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}
